### ISSUE TITLE
V2 Driver can't use id filters but engine passes them anyway

### ISSUE THREAD

**Issue by CharliePoole** [nunit/nunit-console#640]
This doesn't impact the console runner or VS adapter, which don't use id filters. In a GUI, they are the most natural thing to use. 

If the GUI knew when it was running V2 tests, it could pass a different filter. However, it doesn't know that and we wouldn't want each runner to have to deal with this separately.

The problem is exacerbated by the fact that the exception generated is swallowed whole by `MasterTestRunner`, which uses a `when` clause to only handle certain exceptions. Other exceptions are allowed to continue. That would be OK except that the method where this happens is run on a worker thread so the exception goes unhandled.

I'll fix this by first handling the exception and reporting it. Beyond that, we need to decide if the engine should do a filter conversion or the V2 driver should do it. Currently, the V2 driver converts other filters, but not the id filter.

It would also be possible to upgrade the legacy V2 code to directly handle id filters, in which case that feature could be used by a new release of the V2 driver.

I'm assigning this to myself for the purpose of at least reporting the exception. Some discussion on how to take it further is needed to take the next steps.

**Comment by CharliePoole** [comment]
@nunit/engine-team I've made a fix for this in the TestCentric project, to the extent that I"m reporting the error rather than failing silently. It would be cool if some of you would take a look at that PR (TestCentric/testcentric-gui#354) to see if it makes sense to port it here.

For a more final fix, either the V2 adapter or the V2 framework itself should be changed so that an id filter actually works. Then, anyone needing the correct behavior (currently only in the testcentric GUI) would simply upgrade the driver. 

NOTE: if it's fixed in the legacy nunitv2 release, then the v2 driver would just get a package upgrade. If fixed in the driver, then the code to convert from an id filter to a testname filter would be in the driver itself.

I'd prefer not to work on this until I know which approach suits you guys best, since I could then provide the PR to NUnit. OTOH, I have to do something soon, since this is a 1.0 release issue.

**Comment by mikkelbu** [comment]
I don't think you are referring to the right PR, don't you mean TestCentric/testcentric-gui#354 ? Looking at the code I don't see a problem in porting it here (perhaps introducing a helper method to reduce the duplication of creating the attributes).

Regarding the final fix, then I think it should be solved in the framework, if the change is not too complex, but note that I know next to nothing about this part of the code. I see the solution in the adapter as the second option.

**Comment by CharliePoole** [comment]
@mikkelbu That's correct. I'll edit the description.

**Comment by ChrisMaddock** [comment]
I'm not familiar with the v2 driver, does it reference the framework itself? As opposed to how the v3 driver works?

If it was v3, I would have assumed the fix should be in the driver, so the engine will continue to work with all 3.x versions of the framework. Does the v2 driver work differently?

**Comment by CharliePoole** [comment]
The V2 driver uses two V2 assemblies, nunit.core and nunit.core.interfaces, but not the framework. Originally it had its own binary copies but now it uses a package from the nunit legacy project. When I say it could be fixed in the legacy release, I mean in nunit core, which is the part that works with all old versions of the framework.

I guess it makes the most sense to do this in the v2 driver if possible. It depends on whether the core code that needs to be called is all public, which I could only find out by trying.

**Comment by rprouse** [comment]
To me it seems like the driver is the best place to fix this since it is a V3 to V2 interface issue. Am I correct in thinking this would never be an issue with a legacy V2 runner?

**Comment by CharliePoole** [comment]
I'd say that's correct. Legacy runners can use the UniqueName of a test to identify it. That's what the V2 GUI does. I'll take a look at the driver. To me it's really a question of whether enough info is exposed by the V2 framework to allow the conversion. Basically, we need roundtrip conversion between V2 unique names and V3 IDs.

**Comment by CharliePoole** [comment]
This looks doable in the driver. I'll transfer the issue to that project and assign it to myself. Drivers will just need to deal with this filter type. It should not be hard since they have to generate ids anyway.
